src/librustc_data_structures/sync.rs

```diff
@@ -181,46 +199,40 @@ cfg_if! {
             ($($blocks:tt),*) => {
                 // We catch panics here ensuring that all the blocks execute.
                 // This makes behavior consistent with the parallel compiler.
-                let mut panic = None;
+                let panic = ::rustc_data_structures::sync::Lock::new(None);
                 $(
-                    if let Err(p) = ::std::panic::catch_unwind(
-                        ::std::panic::AssertUnwindSafe(|| $blocks)
-                    ) {
-                        if panic.is_none() {
-                            panic = Some(p);
-                        }
-                    }
+                    ::rustc_data_structures::sync::catch(&panic, || $blocks);
                 )*
-                if let Some(panic) = panic {
-                    ::std::panic::resume_unwind(panic);
-                }
+                ::rustc_data_structures::sync::resume(panic);
             }
         }
 
-        pub use std::iter::Iterator as ParallelIterator;
+        use std::iter::{Iterator, IntoIterator, FromIterator};
 
-        pub fn par_iter<T: IntoIterator>(t: T) -> T::IntoIter {
-            t.into_iter()
-        }
-
-        pub fn par_for_each_in<T: IntoIterator>(
+        pub fn par_for_each<T: IntoIterator>(
             t: T,
-            for_each:
-                impl Fn(<<T as IntoIterator>::IntoIter as Iterator>::Item) + Sync + Send
+            mut for_each: impl FnMut(<<T as IntoIterator>::IntoIter as Iterator>::Item),
         ) {
             // We catch panics here ensuring that all the loop iterations execute.
             // This makes behavior consistent with the parallel compiler.
-            let mut panic = None;
+            let panic = Lock::new(None);
             t.into_iter().for_each(|i| {
-                if let Err(p) = catch_unwind(AssertUnwindSafe(|| for_each(i))) {
-                    if panic.is_none() {
-                        panic = Some(p);
-                    }
-                }
+                catch(&panic, || for_each(i));
             });
-            if let Some(panic) = panic {
-                resume_unwind(panic);
-            }
+            resume(panic);
+        }
+
+        pub fn par_map<T: IntoIterator, R, C: FromIterator<R>>(
+            t: T,
+            mut map: impl FnMut(<<T as IntoIterator>::IntoIter as Iterator>::Item) -> R,
+        ) -> C {
+            // We catch panics here ensuring that all the loop iterations execute.
+            let panic = Lock::new(None);
+            let r = t.into_iter().filter_map(|i| {
+                catch(&panic, || map(i))
+            }).collect();
+            resume(panic);
+            r
```

Comment:
I would prefer to avoid the locks in the single-threaded case, as I believe they're not needed and it's confusing as to why they're there. At least we should `// these locks are for consistency` or something like that.

Reply:
`Lock` is just a `RefCell` in the single-threaded case.

Reply:
Good point! Yeah, doesn't matter much then.